lldb/include/lldb/Core/DILAST.h

```diff
@@ -0,0 +1,459 @@
+//===-- DILAST.h ------------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_CORE_DILAST_H
+#define LLDB_CORE_DILAST_H
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "lldb/Core/ValueObject.h"
+#include "lldb/Symbol/Type.h"
+#include "lldb/Symbol/TypeList.h"
+#include "lldb/Target/LanguageRuntime.h"
+#include "lldb/Utility/ConstString.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Basic/TokenKinds.h"
+#include "llvm/ADT/APFloat.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/Support/Casting.h"
+
+namespace lldb_private {
+
+namespace DIL {
+
+/// Struct to hold information about member fields. Used by the parser for the
+/// Data Inspection Language (DIL).
+struct MemberInfo {
+  std::optional<std::string> name;
+  CompilerType type;
+  std::optional<uint32_t> bitfield_size_in_bits;
+  bool is_synthetic;
+  bool is_dynamic;
+  lldb::ValueObjectSP val_obj_sp;
+};
+
+/// Get the appropriate ValueObjectSP, consulting the use_dynamic and
+/// use_synthetic options passed, acquiring the process & target locks if
+/// appropriate.
+lldb::ValueObjectSP GetDynamicOrSyntheticValue(
+    lldb::ValueObjectSP valobj_sp,
+    lldb::DynamicValueType use_dynamic = lldb::eNoDynamicValues,
+    bool use_synthetic = false);
+
+/// The various types DIL AST nodes (used by the DIL parser).
+enum class NodeKind {
+  eErrorNode,
+  eScalarLiteralNode,
+  eStringLiteralNode,
+  eIdentifierNode,
+  eCStyleCastNode,
+  eMemberOfNode,
+  eArraySubscriptNode,
+  eUnaryOpNode,
+  eSmartPtrToPtrDecayNode
+};
+
+/// The C-Style casts for type promotion allowed by DIL.
+enum class TypePromotionCastKind {
+  eArithmetic,
+  ePointer,
+};
+
+/// The Unary operators recognized by DIL.
+enum class UnaryOpKind {
+  AddrOf, // "&"
+  Deref,  // "*"
+  Minus,  // "-"
+};
+
+/// Given a string representing a type, returns the CompilerType corresponding
+/// to the named type, if it exists.
+CompilerType
+ResolveTypeByName(const std::string &name,
+                  std::shared_ptr<ExecutionContextScope> ctx_scope);
+
+/// Class used to store & manipulate information about identifiers.
+class IdentifierInfo {
+private:
+  using MemberPath = std::vector<uint32_t>;
+  using IdentifierInfoUP = std::unique_ptr<IdentifierInfo>;
+
+public:
+  enum class Kind {
+    eValue,
+    eContextArg,
+    eMemberPath,
+    eThisKeyword,
+  };
+
+  static IdentifierInfoUP FromValue(lldb::ValueObjectSP value_sp) {
```

Comment:
Just to make sure I'm not misunderstanding you: You would like me to change the parameter for the FromValue function to be a ValueObject& rather than a ValueObjectSP?

Reply:
Yes, that was they idea, particularly if the callers already know they are dealing with a valid object. I want to avoid needing to check the pointer validity at every level in the stack.

Reply:
I went back and looked at this a little closer; when the user requests a nonexistent variable name (makes a mistake), this function can be passed an empty ValueObjectSP; the parser checks to see if the IdentifierInfo object that gets returned passes the IsValid check, and that's when the parser either continues parsing or tells the user they've made an invalid request. So no, I can't really change the parameter here on the assumption that the function always gets a valid object.

Reply:
Hmm... this ties in to the IsValid discussion we had on a different object (DILMemberInfo). Since all of the factory functions creating these objects return a pointer, I think that having an extra IsValid() method is unnecessary and/or confusing as the (in)validity of the result could be indicated by the (in)validity of the pointer.
That I think would also make it natural to move this check to the caller, as it could do something like:
```
if (value_sp) return IdentifierInfo::FromValue(value_sp);
return nullptr;
```

Reply:
I didn't look further into the code to see if you actually treat ValueObjects this way, but just to make sure this is clear, IsValid is a necessary but not sufficient check to see whether you can use an ValueObject.  You also have to check GetError().Success().  If you think about the case of a variable request for a variable that is defined in the block containing the PC, but is optimized out at the current PC.  That is a valid SBValue, since it carries a useful error and might have a value after you step again.  But it doesn't have a value you can use at this stop.

Reply:
Ack.